Первая итерация 7-го ТЗ.

nev-ivan opened the pull request:
ветка для 7-го ТЗ

Commit:
Первая итерация 7-го ТЗ.

Commit:
Первая итерация 7-го ТЗ.

Commit:
Первая итерация 7-го ТЗ.

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,246 @@
+
+public class FileBackedTaskManager extends InMemoryTaskManager {
+    private final File file;
+    String firstLine = "id,type,name,status,description,epic";
```
Сделай тогда уже это константой, private final static, шапка файла не будет меняться (только в случае если мы расширим состав полей)

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,246 @@
+    public static FileBackedTaskManager loadFromFile(File file) {
+        final FileBackedTaskManager taskManager = new FileBackedTaskManager(file);
+        int counterId = 0;
+        boolean notBlank = true;
```
Флаг по сути лишний, он даже никак не меняется, можно просто пропустить итерацию, если строка пустая

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,246 @@
+                    task.setStatus(TaskStatus.IN_PROGRESS);
+                } else if (taskInArray[3].equals("DONE")) {
+                    task.setStatus(TaskStatus.DONE);
+                }
+                break;
+            default:
+                return null;
+        }
+        return task;
+    }
+
+    private void addTask(Task task) {
+        if (task != null) {
+            switch (task.getType()) {
+                case TASK -> taskMap.put(task.getId(), task);
+                case SUBTASK -> subTaskMap.put(task.getId(), (SubTask) task);
+                case EPIC -> epicMap.put(task.getId(), (EpicTask) task);
+            }
+        }
+    }
+
+    private List<Integer> historyFromString(String line) {
+        if (line == null) {
+            return null;
+        }
+        List<Integer> tasksId = new ArrayList<>();
+        String[] idInStringArray = line.split(",");
+        for (String s : idInStringArray) {
+            tasksId.add(Integer.parseInt(s));
+        }
+        return tasksId;
+    }
+
+    private String historyToString(List<Task> history) {
+        StringBuilder line = new StringBuilder();
+        for (Task task : history) {
+            line.append(task.getId() + ",");
+        }
+        return line.toString();
+    }
```
Методы преобразования из строки/в строку лучше вынести в утилитный класс, чтобы разгрузить менеджер.

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,246 @@
+        }
+    }
+
+    private Task taskFromString(String line) {
+        String[] taskInArray = line.split(",");
+        Task task;
+        switch (taskInArray[1]) {
+            case "SUBTASK":
+                task = new SubTask(Integer.parseInt(taskInArray[0]), taskInArray[2].trim(), taskInArray[4].trim(),
+                        Integer.parseInt(taskInArray[5]));
+                if (taskInArray[3].equals("IN_PROGRESS")) {
+                    task.setStatus(TaskStatus.IN_PROGRESS);
+                } else if (taskInArray[3].equals("DONE")) {
+                    task.setStatus(TaskStatus.DONE);
+                }
+                break;
+            case "TASK":
+                task = new Task(Integer.parseInt(taskInArray[0]), taskInArray[2].trim(), taskInArray[4].trim());
+                if (taskInArray[3].equals("IN_PROGRESS")) {
+                    task.setStatus(TaskStatus.IN_PROGRESS);
+                } else if (taskInArray[3].equals("DONE")) {
+                    task.setStatus(TaskStatus.DONE);
+                }
+                break;
+            case "EPIC":
+                task = new EpicTask(Integer.parseInt(taskInArray[0]), taskInArray[2].trim(), taskInArray[4].trim());
+                if (taskInArray[3].equals("IN_PROGRESS")) {
+                    task.setStatus(TaskStatus.IN_PROGRESS);
+                } else if (taskInArray[3].equals("DONE")) {
+                    task.setStatus(TaskStatus.DONE);
+                }
+                break;
+            default:
+                return null;
+        }
+        return task;
+    }
```
Значения полей лучше распарсить до условий и переиспользовать их, чтобы улучшить читаемость кода и не дублировать `taskInArray[1]` , `taskInArray[2]` и тд.

Так же не понимаю смысла условий со статусами, это енам константы и они записываются в файл в виде строкового алиаса, который можно так же преобразовать в енам через `TaskStatus.valueOf()`

nev-ivan replied:
Если честно, я просто забыл про теорию Enum, метод valueOf()

Ivanov-s-tmn commented on src/manager/FileBackedTaskManager.java:
```diff
@@ -0,0 +1,246 @@
+        if (task != null) {
+            switch (task.getType()) {
+                case TASK -> taskMap.put(task.getId(), task);
+                case SUBTASK -> subTaskMap.put(task.getId(), (SubTask) task);
```
Сабтаски еще нужно не забыть добавить в список внутри эпика

Ivanov-s-tmn commented on src/task/Task.java:
```diff
@@ -7,13 +7,17 @@ public class Task {
+
     protected TaskStatus status;
 
+    protected TaskType type;
```
Тип не нужен как поле, можно обойтись и одним геттером, который будет в зависимости от класса возвращать нужный тип

nev-ivan replied:
Реализовал путем переопределения метода в каждом типе задач, если такая реализация не оптимальна, готов переделать на ваш вариант.

Ivanov-s-tmn replied:
Нет, все ок, именно это я имел в виду)

Ivanov-s-tmn commented on src/task/SubTask.java:
```diff
@@ -7,16 +7,23 @@ public class SubTask extends Task {
     public SubTask(String name, String taskDescription, int idOfEpic) {
         super(name, taskDescription);
         this.idOfEpic = idOfEpic;
+        status = TaskStatus.NEW;
```
Статус инициализируется в конструкторе T`ask`

Ivanov-s-tmn commented on test/manager/FileBackedTaskManagerTest.java:
```diff
@@ -0,0 +1,86 @@
+        task4 = new EpicTask(4, "44", "444");
+        task5 = new SubTask(5, "55", "555", 4);
+        file = File.createTempFile("task", ".csv");
+        taskManager = FileBackedTaskManager.loadFromFile(file);
```
В данном случае можно создать и через `new FileBackedTaskManager(file)`, создание через loadFromFile() можно оставить для теста и проверок восстановления файлового менеджера из файла

Ivanov-s-tmn commented on test/manager/FileBackedTaskManagerTest.java:
```diff
@@ -0,0 +1,86 @@
+    @Test
+    void saveAndLoadFromFileTest() {
+        FileBackedTaskManager taskManagerTest = FileBackedTaskManager.loadFromFile(file);
+        assertEquals(taskManager.taskMap, taskManagerTest.taskMap, "Функция работает некорректно.");
+        assertEquals(3, taskManagerTest.taskMap.size(), "Некорректное значение");
+        assertEquals(1, taskManagerTest.subTaskMap.size(), "Некорректное значение");
+        assertEquals(1, taskManagerTest.epicMap.size(), "Некорректное значение");
```
Сравнивай списки задач файлового менеджера созданного в `beforeEach()` методе и в рамках этого теста (до и после восстановления состояния из файла

`assertEquals(taskManager.getTasks(), taskManagerTest.getTasks());` они должны быть идентичны. Аналогично и с остальными списками задач и с историей

Commit:
Вторая итерация 7-го ТЗ.
- Создать новый утилитный класс для методов преобразования задач и истории
- Оредактировать метод taskToString()
- Изменить работу метода addTask(), id сабтаска добавляется в список сабтасков эпика
- Убрать поле type (Task, Subtask, Epic)
- Изменить работу метода saveAndLoadFromFileTest()

Commit:
Вторая итерация 7-го ТЗ.
- Создать новый утилитный класс для методов преобразования задач и истории
- Оредактировать метод taskToString()
- Изменить работу метода addTask(), id сабтаска добавляется в список сабтасков эпика
- Убрать поле type (Task, Subtask, Epic)
- Изменить работу метода saveAndLoadFromFileTest()
